### ISSUE TITLE
Unexpected re-render with old data

### ISSUE THREAD

**Issue by brumm** [ghengeveld/react-async#8]
Hey,

TL;DR: I'm using `Async.Resolved` to load some async data and unfortunately have to perform some side-effects that will determine whether some action later will error or not.

Here's a test case: https://codesandbox.io/s/vvp32m6kr3
Initial render is fine. Then, click on `set duration 5000ms` and watch the console.

Currently, the whole subtree underneath `Async` re-renders when it receives new props, causing the child-function of `Async.Resolved` to be called again, even though `data` is the old value, and `isLoading` is false.

In my case this causes react-intl to complain about a mismatch between the (old) provided localeData and the newly set locale.

My full example:

```js
export class I18nProvider extends React.PureComponent {
  render() {
    const { children } = this.props
    return (
      <Connect mapStateToProps={({ user: { locale } }) => locale}>
        {locale => (
          <Async promiseFn={loadLocale} watch={locale}>
            <Async.Resolved>
              {([localeData, messages]) => {
                console.log({ locale, localeData, messages })

                addLocaleData(localeData)

                return (
                  <IntlProvider locale={locale} messages={messages}>
                    {children}
                  </IntlProvider>
                )
              }}
            </Async.Resolved>
          </Async>
        )}
      </Connect>
    )
  }
}
```

**Comment by ghengeveld** [comment]
Hey Philipp,

Sorry for the late reply and thanks for using react-async, I hope you like it.

In general you should not be triggering side-effects from within a React render function. It goes against the React principles. What you should probably be using is the `onResolve` callback. 

I'm aware that the whole subtree will rerender when any part of Async's state changes. I'm considering adding a `subscription` prop similar to the one in react-final-form to allow more fine grained control over rerenders. What do you think?

**Comment by brumm** [comment]
Hey Gert,

Yeah, I'm aware about the side-effects. Unfortunately I'm forced to do this for now, because react-intl spews errors if the locale set via the locale prop and the `localeData` passed to `addLocaleData` doesn't match. Its really awkward to use.

`onResolve` sounded like exactly what I needed, but `onResolve` gets called _after_ the children re-render. I'm not sure whether that's intentional.
Demonstrated here: https://codesandbox.io/s/4xv4xz6860?expanddevtools=1&view=editor

I'm not familiar with react-final-form's subscription. From a user perspective I'd expect (especially when using the `watch` prop) that by default the component does not re-render unless the actual promise value changes. I would not expect (but accept I guess) having to specify which updates I care about via some subscription prop :)

**Comment by ghengeveld** [comment]
Hi Philipp,

Are you still in need of this feature? I prefer not to overcomplicate the package. I think you can do something like this to avoid calling you function on every render:

```js
class Watch extends React.Component {
  componentDidMount() {
    this.props.onChange()
  }
  componentDidUpdate(prevProps) {
    if (prevProps.watch !== this.props.watch) this.props.onChange()
  }
  render() {
    return this.props.children || null
  }
}
```

Then wrap your side-effect in a `<Watch onChange={doSideEffect}>` instance inside `<Async.Resolved>` to get what you're looking for. You can pass a `watch` prop to Watch in order to have it invoke the onChange callback again, but probably you don't even need it.
